Improved LIS and added Rabbit solution (#1077)
cpinitiative/usaco-guide: content/4_Gold/LIS.mdx

In this case, $L_8$ would be $[2, 5, 15]$. Some example sequences satisfying
these are $[2]$, $[2, 5]$, and $[2, 5, 15]$.

**Lemma 1:** $L_i$ forms a strictly increasing sequence.

**Proof:** Assume for a contradiction that for some $j$, we have
$L_i[j - 1] \geq L_i[j]$. Let $a_0, a_1, \dots, a_{j}$ be any increasing
subsequence of length $j + 1$ ending on $L_i[j]$ (so $a_j = L_i[j]$). Notice
that $a_0, a_1, \dots, a_{j - 1}$ is an increasing subsequence of length $j$.
Thus, $L_i[j]$ has to be greater than $L_i[j - 1]$. From that, we can write this
equation: $L_i[j - 1] \leq a_{j-1} < a_j = L_i[j]$, which is exactly what we
wanted.

**Lemma 2:** The length of the LIS ending on $A[i + 1]$ is equal to the least
index $j$ (with 0-indexing) such that $L_i[j] \geq A[i + 1]$.

**Proof:** Let $j$ be defined as it is in the statement. First of all, since
$A[i + 1] > L_i[j - 1]$, there is an increasing sequence of length $j$ ending on
$A[i + 1]$, as we can append $A[i + 1]$ to the end of the sequence of length
$j - 1$. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
all $k \geq j$. This means that there cannot be a LIS ending at $A[i + 1]$
longer than $j$.

> 

Reviewer: Shift indicies

**Lemma 3:** At most 1 element differs between $L_i$ and $L_{i + 1}$.

**Proof:** Let $j$ be defined as it was in Lemma 2. We need to set
$L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$.
$L_{i + 1}[k] = L_i[k]$ for all $k \neq j$, though, since $A[i + 1] > L_i[k]$
for all $k < j$ and there are no increasing sequences ending on $A[i + 1]$ for

> $k > j$.

Reviewer: Does this need index shifting too?

Author: Don't think so...

elements of $A$ with an increasing sequence of length $j$ ending on it (or
$\infty$ if there is no such element).

For example, let our array be `2 18 7 20 18 5 18 15 13 19 9`.

> In this case, $L_8$ would be `[2, 5, 15]`, some example sequences satisfying

Reviewer 2: ```suggestion
In this case, $L_8$ would be `[2, 5, 15]`. Some example sequences satisfying
```

> In this case, $L_8$ would be `[2, 5, 15]`. Some example sequences satisfying

Author: Actually yeah, L_8[j] is the minimal ending element of a sequence that has length j + 1 instead of length j.

> these being `[2]`, `[2, 5]`, and `[2, 5, 15]`.

Reviewer 2: ```suggestion
these are `[2]`, `[2, 5]`, and `[2, 5, 15]`.
```

**Lemma 1:** $L_i$ forms a strictly increasing sequence.

**Proof:** Assume for a contradiction that for some $j$, we have

> $L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$th

Reviewer 2: ```suggestion
$L_i[j] \geq L_i[j + 1]$. Define $a_i$ ($0 \leq i \leq j$) to be the $i$-th
```

element of any increasing subsequence that ends with $L_i[j + 1]$. We can have

> an increasing subsequence on length $j$ that ends on $a_{j - 1}$ which is less

Reviewer 2: ```suggestion
an increasing subsequence of length $j$ that ends on $a_{j - 1}$, which is less
```

> than $a_j = L_i[j + 1]$, so there is a contradiction.

Author: >  (Note, $L_i[j]$ may be less than $a_{j-1}$ for all sequences $a$ of length $j+1$.)

Wait, what does this mean?

> than $a_j = L_i[j + 1]$, so there is a contradiction.

Reviewer: Probably unnecessary, but it basically means: the optimal answer for a subsequence of length j isn't necessary the first $j$ terms of an optimal subsequence of length j+1

**Lemma 2:** The length of the LIS ending on $A[i + 1]$ is equal to the least
index $j$ (with 0-indexing) such that $L_i[j] \geq A[i + 1]$.

> **Proof:** Let $j$ be defined as it is in the statement. First of all, since

Reviewer: ```suggestion
**Proof:** Since
```

Author: You sure? I feel like that makes it a bit more confusing.

Reviewer: It's fine as is, but seems redundant

$A[i + 1] > L_i[j - 1]$, there is an increasing sequence of length $j$ ending on

> $A[i + 1]$ as we can append $A[i + 1]$ to the end of the sequence of length

Reviewer 2: ```suggestion
$A[i + 1]$, as we can append $A[i + 1]$ to the end of the sequence of length
```

> $j - 1. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for

Reviewer 2: ```suggestion
$j - 1$. By Lemma 1, $L_i$ is strictly increasing, so $L_i[k] \geq A[i + 1]$ for
```

> all $k \geq j$. This means that the length of the LIS is $j$.

Reviewer: ```suggestion
all $k \geq j$. This means that there cannot be a LIS ending at $A[i+1]$ longer than $j$.
```

**Lemma 3:** At most 1 element differs between $L_i$ and $L_{i + 1}$.

> **Proof:** Let $j$ be defined as it was in the previous lemma. We need to set

Reviewer: ```suggestion
**Proof:** Let $j$ be defined as in Lemma 2. We need to set
```

> $L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$. We don't update
> anything else though, since $A[i + 1] > L_i[k]$ for all $k < j$ and there are no
> increasing sequences ending on $A[i + 1]$ of length greater than $j$.

Reviewer: ```suggestion
$L_{i + 1}[j]$ to be $A[i + 1]$ since $A[i + 1] \leq L_i[j]$. $L_{i + 1}[k] = L_i[k]$
for all $k \neq j$, though, since $A[i + 1] > L_i[k]$ for all $k < j$ and there are no
increasing sequences ending on $A[i + 1]$ for $k > j$.
```

Maybe this helps?

To find and update the described $j$ in $\mathcal{O}(\log N)$ time, we can use a

> list with binary search, or we can use a sorted set (as demonstrated in the

Reviewer: I think "ordered set" is more accurate

**Lemma 1:** $L_i$ forms a strictly increasing sequence.

**Proof:** Assume for a contradiction that for some $j$, we have
$L_i[j] \geq L_i[j + 1]$. Let $a_0, a_1, \dots, a_j$ be any increasing subsequence
of length $j+1$ ending on $L_i[j + 1]$ (so $a_j=L_i[j + 1]$). Notice that
$a_0, a_1, \dots, a_{j-1}$ is an increasing subsequence of length $j$.
Thus, $L_i[j]$ cannot be greater than $a_{j-1}$.
From that, we can write this equation: $L_i[j] \leq a_{j-1} < a_j = L_i[j+1]$,

> which is exactly what we wanted.

Reviewer: Uhh, is this still accurate per the new definition of L_i[j]?

I believe indicies should be shifted by 1

Author: Oh yeah, I'll change that.

	vector<int> dp;
	for (int i : a) {
		int pos = lower_bound(dp.begin(), dp.end(), i) - dp.begin();
		if (pos == dp.size()) {  // we can have a new, longer IS!
			dp.push_back(i);
		} else {  // oh ok, at least we can make the ending element smaller
			dp[pos] = i;

> 		}

Reviewer: Fine as is, no?

Author: I feel like just adding braces make it look better. I can still remove them if you want.

Reviewer: Oh, I personally think braces for 1-liners looks messy. Generally I leave stuff like this to the author of the code though

Reviewer 3: My two cents: leave the braces in so control flow is easier to understand.

	vector<int> dp;
	for (int i : a) {
		int pos = lower_bound(dp.begin(), dp.end(), i) - dp.begin();
		if (pos == dp.size()) {  // we can have a new, longer increasing subsequence!
			dp.push_back(i);
		} else {  // oh ok, at least we can make the ending element smaller
			dp[pos] = i;

> 		}

Reviewer: Leave this as is.

<JavaSection>

```java
public static int findLis(int[] a) {
	ArrayList<Integer> dp = new ArrayList<Integer>();
	for (int i : a) {
		int pos = Collections.binarySearch(dp, i);
		pos = pos < 0 ? Math.abs(pos + 1) : pos;
		if (pos == dp.size()) {  // we can have a new, longer increasing subsequence!
			dp.add(i);
		} else {  // oh ok, at least we can make the ending element smaller
			dp.set(pos, i);
		}
	}
	return dp.size();
}
```

</JavaSection>

<PySection>

```py

> # list from typing module

Reviewer 2: Do you think it's better to explicitly import these modules?

Author: C++ and Java didn't, so I'm not sure.

Reviewer 2: Ok then it's fine

<JavaSection>

```java
public static int findLis(int[] a) {
	ArrayList<Integer> dp = new ArrayList<Integer>();
	for (int i : a) {
		int pos = Collections.binarySearch(dp, i);
		if (pos < 0) {
			pos = Math.abs(pos + 1);
		}

> 		if (pos == dp.size()) {  // we can have a new, longer IS!

Reviewer 2: Rather say "increasing subsequence" instead of "IS"

			dp.add(i);
		} else {  // oh ok, at least we can make the ending element smaller
			dp.set(pos, i);

> 		}

Reviewer: Seems unnecessarily long compared to the 3 concise lines D:

Reviewer 3: They're equally concise, though, because the code is the same. One just makes control flow more obvious.

Reviewer 2: It's a single statement though. I'd argue that making the code more vertically concise is better for readability because it's more compact and elegant

Author: I mean with monitors as big as they are, being vertically concise shouldn't be that big of a concern (also I compacted the `Math.abs` part).

Reviewer 3: > more compact and elegant

No, the former code is not, in fact, more elegant.

Reviewer 2: How so?

Reviewer: > control flow more obvious

Yes, in the original way you have a series of (condition -> statement). This exactly contributes to why concise is cleaner.

The alternative requires you to connect condition & statement between multiple lines. Of course this would be preferred for a set of substantial, multiline statement, but not when it is a sleek one-liner.

Reviewer: In the end settling with the choice of the original author is just more respectful, especially when there is no significant reason to change.

Reviewer 3: By virtue of the semantics being the exact same. The indentation present in the latter, however, makes the control flow easier to digest—that is the point of indentation. Braces are a good practice anyway and doesn't make it any less "elegant" or readable.

	}
	return dp.size();
}
```

</JavaSection>

<PySection>

> # list from typing module

Reviewer 2: Wrap the code in triple back ticks

Author: Oh yeah, forgot about that.

<JavaSection>

```java
public static int findLis(int[] a) {
	ArrayList<Integer> dp = new ArrayList<Integer>();
	for (int i : a) {
		int pos = Collections.binarySearch(dp, i);
		pos = pos < 0 ? Math.abs(pos + 1) : pos;
		if (pos == dp.size()) {  // we can have a new, longer increasing subsequence!
			dp.add(i);
		} else {  // oh ok, at least we can make the ending element smaller
			dp.set(pos, i);

> 		}

Reviewer: Leave as is

(I don't believe there is a compelling reason to change to camelcase, right?)

Author: The convention for Java is camelCase. It's always good to be consistent.
